CLDC-1852 add weekly rent period and reorder options

arfacamble opened the pull request:
https://digital.dclg.gov.uk/jira/browse/CLDC-1852

1. add in a new option for the period question and reorder existing answer options
2. ensure that the lettings log calculates values correctly that relate to the period

|<img width="985" alt="image" src="https://user-images.githubusercontent.com/51094020/229497064-660fd720-a4ee-448f-9e2c-45c62c1c1264.png">
|-

|<img width="800" alt="image" src="https://user-images.githubusercontent.com/51094020/230343432-d5be672a-9fc6-490b-8657-a642493e12fe.png">
|-

arfacamble committed:
add in a new option for the period question and reorder existing answer options

arfacamble committed:
ensure that the lettings log calculates values correctly that relate to the period

SamSeed-Softwire reviewed (commented):
The stuff you've written all looks fine.

Are there other places in the codebase however which need to reference the new answer option? E.g. the following look like they might need it:
- `app/models/form/lettings/pages/care_home_weekly.rb`
- `app/models/form/lettings/pages/rent_weekly.rb`
- `app/models/form/lettings/questions/tshortfall.rb`
- `config/forms/2022_2023.json`, assuming it's relevant to 22/23

Is it worth adding some tests for the new answer option in `spec/models/lettings_log_spec.rb` as well?

arfacamble committed:
ensure that depends on conditions related to period take account of the new option
relatedly create methods on LettingsLog to allow improved readability in these depends on conditions
refactor depends on conditions with these new method

arfacamble committed:
update the 22_23 json form to use the new option as well

arfacamble committed:
update tests to cover deriving variables that depend on period
update test fixture to allow this

arfacamble committed:
refactor some depends on and amend LEttingsLog method to improve readability

arfacamble committed:
remove forgotten focus in spec file found in linting check

arfacamble committed:
amend context descriptino in new tests

SamSeed-Softwire commented on app/models/lettings_log.rb:
```diff
@@ -379,6 +379,10 @@ def receives_uc_with_housing_element_excl_housing_benefit?
     hb == 6
   end
 
+  def receives_any_housing_benefit?
```
Could this new method also be used to refactor `receives_housing_related_benefits?`? Or did you even need to create this method and could you have just used `receives_housing_related_benefits?` instead?

arfacamble replied:
just use that! good catch

SamSeed-Softwire commented on spec/fixtures/forms/2021_2022.json:
Is the new answer option necessary here, given this file is for 2021/22?

arfacamble replied:
this is a test fixture. If I don't put it here I have to rewrite the whole lettings_log_spec and that would really stretch the scope of this ticket!

SamSeed-Softwire replied:
Ah gotcha

SamSeed-Softwire commented on app/models/form/lettings/pages/rent_weekly.rb:
```diff
@@ -4,30 +4,8 @@ def initialize(id, hsh, subsection)
-      { "period" => 8, "household_charge" => nil, "is_carehome" => nil },
-      { "period" => 9, "household_charge" => 0, "is_carehome" => nil },
-      { "period" => 9, "household_charge" => nil, "is_carehome" => nil },
+      { "rent_and_charges_paid_weekly?" => true, "household_charge" => 0, "is_carehome?" => false },
```
There are various other places that `"is_carehome" => nil or 0` could be replaced with `"is_carehome?" => false`, for example in `app/models/form/lettings/pages/rent_4_weekly.rb`. I'd recommend either doing that as part of this ticket, or raising a new tech debt ticket.

arfacamble replied:
Not sure, I don't want to change half the conditions in rent_4_weekly, so then I change them all with some new methods, and then I have to search to for everywhere that household_charge is nil or 0 and everywhere period is 3 and then write new methods for the other conditions in those depends on, and so on and so on.
Where do I cut the scope of the ticket off?

arfacamble replied:
ok I relented on this in a big way

SamSeed-Softwire replied:
> ok I relented on this in a big way

What do you mean?

arfacamble replied:
I made the changes

arfacamble committed:
refactor a couple more depends on for readability

SamSeed-Softwire reviewed (approved):
The time has come!
